### ISSUE TITLE
Certain Turing machine configurations cause site to crash

### ISSUE THREAD

**Issue by tdib** [automatarium/automatarium#465]
### Description of the bug

On certain configurations of Turing machines, opening the testing lab causes the site to freeze and crash.

### To reproduce

Unsure exactly what is causing this behaviour, but constructing a machine like the below causes the crashing. Removing any of the transitions in this machine causes the crashing to stop.

<img width="422" alt="image" src="https://github.com/automatarium/automatarium/assets/79500311/c4de866a-d4e4-4af2-93d7-8fee83d8ebaa">

### Expected behavior

No crash :)

**Comment by anioncat** [comment]
Hmm this is similar to #410. Since there's no final state and and there is a cycle of lambdas from the source node the simulation would just search forever (since when you open the testing lab fresh you have the input of just lambda, so the tape is full of infinite lambdas). I think the proposed solution is to keep a counter on the search and ask the user "I expanded 10000 nodes, continue?" or similar.

**Comment by tdib** [comment]
hmm right that'd make sense, I initially had a final state, although that was also crashing, I tried to reduce this to the minimum number of states while still encountering this - wondering if the computation has to happen instantly? I was just wanting to step through one by one but it seems to apply the entire computation as soon as lab is open?

**Comment by anioncat** [comment]
> hmm right that'd make sense, I initially had a final state, although that was also crashing, I tried to reduce this to the minimum number of states while still encountering this - wondering if the computation has to happen instantly? I was just wanting to step through one by one but it seems to apply the entire computation as soon as lab is open?

No I don't think it needs to happen immediately either. I think that it should operate like the stepping lab which would likely solve both the infinite looping problem, since it's user directed, and the input finishing prematurely on TMs when a final state has an out transition, as that should just run a get successors each step. Kinda wishing I looked more closely at this now, but I'll wait to see what James wants to do with the project.

**Comment by s3897720** [comment]
I do get a similar error with a PDA:
![image](https://github.com/automatarium/automatarium/assets/103556103/2877fedb-1e4a-4954-8449-879e91e09fce)
So the "Killer Machines" behaviour isn't unique to TMs it seems...

**Comment by ire4ever1190** [comment]
Don't really take much. Think its just down to the design where it requires the machine to finish before showing the steps.
e.g. This PDA causes the problem since it can keep adding to the stack
![image](https://github.com/automatarium/automatarium/assets/19339842/8faa052d-b90d-4f84-9a60-0bca3a5328bf)
Maybe making the simulation code lazily evaluate/have step limit would solve it?

**Comment by tdib** [comment]
This was one of the issues we faced when implementing FSA (and the code was rewritten since I believe) - I originally used a step limit of 100 but that wasn't a good idea because it's technically possible to construct a valid machine that would break by this logic (e.g. a massive machine that goes like 150 states of lambdas).

I don't think lazy evaluation is what we want here because the skip to end button will need to evaluate everything and will still break it since there is no outcome.

Not really sure how to approach this, I thought we had something to do with transitive states and knowing what is reachable, but that might also only be applicable for FSA

**Comment by ire4ever1190** [comment]
Combo of lazily eval and then check if its ok with the user like JFLAP has could work
![image](https://github.com/automatarium/automatarium/assets/19339842/3d8c38ce-7a55-4574-a50f-8a824e03978a)
Check after 1000, 2000, 4000, etc

**Comment by tdib** [comment]
I know this is basically the premise of the halting problem but would it be possible to detect cycles for something like this? I've always found that JFLAP feature to be a bit gross

**Comment by ire4ever1190** [comment]
Cycles are already detected since the BFS tracks already seen states. The PDA I showed isn't stuck in a cycle since it keeps creating new states (The stack just keeps growing).

Maybe query chatGPT and ask it if it thinks the machine will halt or not 😜

**Comment by tdib** [comment]
Right yeah slight mess up in terminology from me, but I still feel like it could be possible to detect

If ChatGPT could work it out, I'm sure it's possible 😤
![image](https://github.com/automatarium/automatarium/assets/79500311/79720dfd-ff14-4d21-a126-bd1b9d120ce3)

I don't have a good idea of how/what you need to detect, but maybe someone smart can work it out (if it's even possible) :^)

**Comment by tdib** [comment]
Hmm how does knowing all the transitions help to solve this issue? I'd assume that's what's happening under the hood to pick the accepted path, just not being exposed to the user because we want to keep things simple and only show the end state if accepted

**Comment by s3897720** [comment]
Let me try to explain my idea with images. Taking the original killer machine:
![demo](https://github.com/automatarium/automatarium/assets/103556103/62760a9e-9d11-44f2-aa99-e7aadf65964d)

The original machine has 3 paths: 
0,0;R 
1,1;R 
λ,λ;R
If we feed to the simulator the TM with whichever path we just took removed (if we went from qX -> qX), we avoid infinite recursion, but we also force the looking of other paths in the same sim. I.e. if the sim in the case of the PDA I posted, it would at some stage be forced to succeed (if I remember my CT properly) on the string "aba" by default. The idea being to then take the arrows in the image and present them to the user as "choices" if applicable

Drawing this out, there are definite challenges to this design. First of all being that the tape is not considered. If we had a tape [1] [1] [1] [λ] on the original machine, following the diagram means that the machine will incorrectly halt when trying to return to the first '1'. Second is with the turquoise arrows, if this was implemented there would have to be a way to check that we aren't cycling between the two infinitely (because we are back to the same problem)

Maybe I am chasing my own tail and/or trying to overcomplicate the issue?

I guess the alternative is "If transition X goes from qY to qY Z times, then prevent X from being taken again (for now), but show the user that the result of doing X ∞ times is probably un-halting/inconclusive". This sounds way simpler... but still needs the knowledge of all transitions in the machine. I guess you can implement this as above images show. Perhaps that was the idea I had but lost my head when explaining

**Comment by tdib** [comment]
Not sure I fully understand, I don't really get how removing a certain transition would avoid infinite recursion, and I'd expect that to even give incorrect results but I'm sure there's something that hasn't clicked for me yet 😔 While I don't really get it 😓, if you think it'll work I say give it a shot

Thinking about it, the arbitrary limit of 100/1000 visits to the same state(s) could also be an okay solution if the above doesn't work (even though it's a bandaid solution), and since I originally had that logic in there I don't think it should be too hard to put it back in - I don't _like_ it but it's better than trying to whip up a machine and having the whole site just crash on you :(

Also something else to note is that I believe this was happening when I opened the testing lab, not even clicking any buttons

**Comment by s3897720** [comment]
Yeah I experienced the same with the PDA I posted - open test lab -> rip. 

Give me a week (hopefully) to consider my life and I'll get back to you on its feasibility. This will hopefully include code mockups and HD diagrams which should increase clarity. Or I'll go insane. Whichever first I guess

**Comment by LvandoApps** [comment]
Avengers assemble

**Comment by s3897720** [comment]
![Screenshot_20240501_104352](https://github.com/automatarium/automatarium/assets/103556103/54a8982e-f11e-48bc-9fc0-4d92c3606f28)
Haven't been able to think it out yet but something interesting (if we can't solve it in time) is that the program doesn't crash if you are already in the testing lab

**Comment by tdib** [comment]
I'd presume that maybe the computation happens when you press one of the buttons, or when you first open the lab (otherwise the computation would be happening after every change to the automaton which we don't want)

Not sure if we really need to run that computation when opening the testing lab for the first time anyway but that's pretty minor
